### PR TITLE
support rolecert, jwt

### PR DESCRIPTION
**Pull request by kevindiu**
1. Default disable jwk functionality. 
2. Add new debug server and new endpoint /policycache.

### CODE REVIEW

**Comment by kevindiu** [comment]
- [ ] Fix test case
- [ ] Add disable debug server option? (disable debug server by default?)
- [ ] Program comments.

File: service/server.go [hunk 1]
```diff
@@ -76,7 +80,7 @@ var (
 //
 // The health check server is a http.Server instance, which the port number is read from "config.Server.HealthzPort"
 // , and the handler is as follow - Handle HTTP GET request and always return HTTP Status OK (200) response.
-func NewServer(cfg config.Server, h http.Handler) Server {
+func NewServer(cfg config.Server, h http.Handler, dh http.Handler) Server {
```
Comment (kpango) [inline service/server.go]: can you implement functional option params?

Comment (WindzCUHK) [reply service/server.go]: @kevindiu 
something like this?
`func NewServer(cfg config.Server, serverOptions ...Option) Server {`

File: service/server.go [hunk 2]
```diff
@@ -76,36 +82,43 @@ var (
 //
 // The health check server is a http.Server instance, which the port number is read from "config.Server.HealthzPort"
 // , and the handler is as follow - Handle HTTP GET request and always return HTTP Status OK (200) response.
-func NewServer(cfg config.Server, h http.Handler) Server {
+func NewServer(opts ...Option) Server {
```
Comment (WindzCUHK) [inline service/server.go]:
```suggestion
func NewServer(cfg config.Server, serverOpts ...Option) Server {
```

File: handler/debug_handler.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+func (dh *DebugHandler) GetPolicyCache(w http.ResponseWriter, r *http.Request) error {
+	w.WriteHeader(http.StatusOK)
+	w.Header().Set(ContentType, fmt.Sprintf("%s;%s", ApplicationJson, CharsetUTF8))
+	/*
+		e := json.NewEncoder(w)
+		e.SetIndent("", "\t")
+		return e.Encode(dh.authd.GetPolicyCache(r.Context()))
+	*/
+	_, err := fmt.Fprintf(w, "%+s", dh.authd.GetPolicyCache(r.Context()))
+	return err
```
Comment (WindzCUHK) [inline handler/debug_handler.go]: follow the client sidecar style?
https://github.com/yahoojapan/athenz-client-sidecar/blob/master/handler/handlers.go#L67-L79

**Comment by kevindiu** [comment]
It doesn't make sense that debug.New() return a handler. I think it is rather than handler.NewDebug() or debug.NewHandler(). 
But `debug` package name means nothing for me...

**Comment by WindzCUHK** [comment]
> It doesn't make sense that debug.New() return a handler. I think it is rather than handler.NewDebug() or debug.NewHandler().
> But `debug` package name means nothing for me...

Just get an idea from `net/http/httptest`.
maybe `handler/debughandler` as package name, just think that the debug endpoint should not be on the same package

File: handler/debug_handler.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+	// CharsetUTF8 represents a UTF-8 charset for HTTP response "charset=UTF-8"
+	CharsetUTF8 = "charset=UTF-8"
+)
+
+type DebugHandler struct {
+	authd service.Authorizationd
+}
+
+func NewDebugHandler(authd service.Authorizationd) *DebugHandler {
```
Comment (kpango) [inline handler/debug_handler.go]:
```
func NewDebugHandler(authd service.Authorizationd)(func (w http.ResponseWriter, r *http.Request) error){
return func (w http.ResponseWriter, r *http.Request) error {
	w.WriteHeader(http.StatusOK)
	w.Header().Set(ContentType, fmt.Sprintf("%s;%s", ApplicationJson, CharsetUTF8))
	e := json.NewEncoder(w)
	e.SetIndent("", "\t")
	pc := dh.authd.GetPolicyCache(r.Context())
	return e.Encode(pc)
}
}
```

File: router/debug_routers.go [hunk 5]
```diff
@@ -13,15 +17,26 @@ type Route struct {
...
+func NewDebugRoutes(cfg config.Server, a service.Authorizationd) []Route {
 	return []Route{
 		{
 			"GetPolicyCache",
 			[]string{
 				http.MethodGet,
 			},
-			"/policyCache",
-			h.GetPolicyCache,
+			"/debug/policy/cache",
```
Comment (WindzCUHK) [inline router/debug_routers.go]: `"/debug/policy-cache"`?

File: config/config.go [hunk 6]
```diff
@@ -58,6 +58,9 @@ type Server struct {
 	// Port represents the server port.
 	Port int `yaml:"port"`
 
+	// EnableDebug represents if user want to enable debug funcationality.
+	EnableDebug bool `yaml:"enable_debug"`
```
Comment (WindzCUHK) [inline config/config.go]: maybe just check `DebugPort` is zero or not? no need this flag

**Comment by WindzCUHK** [comment]
@kevindiu 
please also help to rename this
`git mv service/authorized_mock.go service/authorizerd_mock.go`

File: config/config.go [hunk 7]
```diff
@@ -64,6 +64,9 @@ type Server struct {
 	// DebugPort represents debug server port.
 	DebugPort int `yaml:"debug_port"`
 
+	// DebugPolicyCachePath represents the policy cache debug path
+	DebugPolicyCachePath string `yaml:"debug_policy_cache_path"`
```
Comment (WindzCUHK) [inline config/config.go]: maybe this is too much for user?
I think it is easier for them to follow the hard coded endpoint.

**Comment by WindzCUHK** [comment]
@kevindiu 
I think it should be `routes`
```bash
git mv router/debug_routers.go router/debug_routes.go
git mv router/debug_routers_test.go router/debug_routes_test.go
```

https://github.com/yahoojapan/athenz-client-sidecar/tree/master/router

**Pull request merged**